src/schema-generation/order-by-and-pagination-augmentation.ts

```diff
@@ -91,12 +109,50 @@ export class OrderByAndPaginationAugmentation {
                     description: `The number of items to include in the result. If omitted, all remaining items will be included (which can cause performance problems on large collections).`,
                 },
             },
+            transformResult: (data: any, args: object, executionOptions: ExecutionOptions) => {
+                if (
+                    options?.firstLimitCheckType !== LimitTypeCheckType.RESOLVER ||
+                    'first' in args
+                ) {
+                    return data;
+                }
+                if (
+                    executionOptions.implicitLimitForRootEntityQueries &&
+                    data.length > executionOptions.implicitLimitForRootEntityQueries
+                ) {
+                    throw new Error(
+                        `Collection is truncated by default to ${executionOptions.implicitLimitForRootEntityQueries} elements but contains more elements than this limit. Specify a limit manually to retrieve all elements of the collection.`,
+                    );
+                }
+                return data;
+            },
             resolve: (sourceNode, args, info) => {
                 let listNode = schemaField.resolve(sourceNode, args, info);
                 let itemVariable = new VariableQueryNode(decapitalize(type.name));
                 let objectNode = this.rootFieldHelper.getRealItemNode(itemVariable, info);
 
-                const maxCount: number | undefined = args[FIRST_ARG];
+                let maxCount: number | undefined = args[FIRST_ARG];
+
+                if (
+                    options?.firstLimitCheckType &&
+                    info?.maxLimitForRootEntityQueries &&
+                    maxCount !== undefined &&
+                    maxCount > info.maxLimitForRootEntityQueries
+                ) {
+                    return new RuntimeErrorQueryNode(
+                        `The requested number of elements via the \"first\" parameter (${maxCount}) exceeds the maximum limit of ${info.maxLimitForRootEntityQueries}.`,
+                        { code: ARGUMENT_OUT_OF_RANGE_ERROR },
+                    );
+                }
+
+                if (
+                    options?.firstLimitCheckType &&
+                    info.implicitLimitForRootEntityQueries &&
+                    maxCount === undefined
+                ) {
+                    maxCount = info.implicitLimitForRootEntityQueries + 1;
```

Comment:
I would set a variable here that says "hey, we need to add a result validator" because this is where this is decided. Makes the conditionals further down simpler